src/internal/storage/fileset/index/writer.go

```diff
@@ -44,30 +43,19 @@ func NewWriter(ctx context.Context, chunks *chunk.Storage, tmpID string) *Writer
 		ctx:    ctx,
 		chunks: chunks,
 		tmpID:  tmpID,
+		closed: make(chan struct{}, 1),
 	}
 }
 
 // WriteIndex writes an index entry.
 func (w *Writer) WriteIndex(idx *Index) error {
-	w.mu.Lock()
-	defer w.mu.Unlock()
-	w.setupLevels()
 	return w.writeIndex(idx, 0)
 }
 
-func (w *Writer) setupLevels() {
-	// Setup the first index level.
-	if w.levels == nil {
-		cw := w.chunks.NewWriter(w.ctx, w.tmpID, w.callback(0), chunk.WithRollingHashConfig(averageBits, 0))
-		w.levels = append(w.levels, &levelWriter{
-			cw:  cw,
-			pbw: pbutil.NewWriter(cw),
-		})
-	}
-}
-
 func (w *Writer) writeIndex(idx *Index, level int) error {
-	l := w.levels[level]
+	w.setupLevel(level)
```

Comment:
Lets set the first index when we create the level rather than in the callback. We would just pass it into the setup and it would be used if the level does not exist. This will allow us to perform the first == last check before closing so we could avoid creating an unnecessary chunk for the last level.